File: artifacts/api-server/src/routes/ranking.ts
// Rutas del ranking: clasificacion global y por modulo de todos los estudiantes
// El ranking muestra puntos, precision, tiempo y mas estadisticas con colores
import { Router } from "express";
import { db } from "@workspace/db";
import { resultadosTable, perfilesTable } from "@workspace/db";
import { eq, desc, sum, avg, count } from "drizzle-orm";
import { requireAuth } from "../lib/auth";

const router = Router();

// GET /ranking - Obtener el ranking global de todos los estudiantes
router.get("/ranking", requireAuth, async (req, res) => {
  try {
    // Agrupa resultados por usuario y calcula estadisticas
    const rankingData = await db
      .select({
        id_usuario: resultadosTable.id_usuario,
        puntos_totales: sum(resultadosTable.puntuacion),
        retos_completados: count(),
        precision_promedio: avg(resultadosTable.precision),
        tiempo_promedio: avg(resultadosTable.tiempo_total),
        respuestas_correctas: sum(resultadosTable.respuestas_correctas),
        respuestas_incorrectas: sum(resultadosTable.respuestas_incorrectas),
      })
      .from(resultadosTable)
      .groupBy(resultadosTable.id_usuario)
      .orderBy(desc(sum(resultadosTable.puntuacion)));

    // Para cada entrada del ranking, agrega informacion del perfil
    const rankingConPerfiles = await Promise.all(
      rankingData.map(async (entry, index) => {
        const [perfil] = await db
          .select({
            id: perfilesTable.id,
            nombre: perfilesTable.nombre,
            usuario: perfilesTable.usuario,
            rol: perfilesTable.rol,
            grado_bachillerato: perfilesTable.grado_bachillerato,
            avatar_url: perfilesTable.avatar_url,
            puntos_totales: perfilesTable.puntos_totales,

            
            retos_completados: perfilesTable.retos_completados,
            racha_dias: perfilesTable.racha_dias,
            ultimo_acceso: perfilesTable.ultimo_acceso,
            creado_en: perfilesTable.creado_en,
            ultimo_nivel_intento: perfilesTable.ultimo_nivel_intento,
            mejor_puntaje_por_modulo: perfilesTable.mejor_puntaje_por_modulo,
          })
          .from(perfilesTable)
          .where(eq(perfilesTable.id, entry.id_usuario));

        // Solo incluye estudiantes en el ranking
        if (!perfil || perfil.rol !== "estudiante") return null;

        return {
          posicion: index + 1,
          usuario_id: perfil.id,
          nombre: perfil.nombre ?? perfil.usuario,
          usuario: perfil.usuario,
          grado_bachillerato: perfil.grado_bachillerato,
          rol: perfil.rol,
          avatar_url: perfil.avatar_url,
          racha_dias: perfil.racha_dias,
          puntos_totales: Number(entry.puntos_totales || 0),
          retos_completados: Number(entry.retos_completados || 0),
          precision_promedio: Number(entry.precision_promedio || 0),
          tiempo_promedio: Number(entry.tiempo_promedio || 0),
          respuestas_correctas: Number(entry.respuestas_correctas || 0),
          respuestas_incorrectas: Number(entry.respuestas_incorrectas || 0),
        };
      })
    );

    // Filtra nulos (docentes) y renumera posiciones
    const rankingFiltrado = rankingConPerfiles
      .filter(Boolean)
      .map((entry, index) => ({ ...entry!, posicion: index + 1 }));

    res.json(rankingFiltrado);
  } catch (err) {
    req.log.error({ err }, "Error al obtener ranking");
    res.status(500).json({ error: "server_error", message: "Error interno" });
  }
});

// GET /ranking/module/:moduleId - Ranking filtrado por modulo especifico
router.get("/ranking/module/:moduleId", requireAuth, async (req, res) => {
  try {
    const moduleId = Number(req.params.moduleId);

    // Nota: para filtrar por modulo necesitamos hacer join con retos
    // Por ahora retorna el ranking global (se puede refinar con join a futuro)
    const rankingData = await db
      .select({
        id_usuario: resultadosTable.id_usuario,
        puntos_totales: sum(resultadosTable.puntuacion),
        retos_completados: count(),
        precision_promedio: avg(resultadosTable.precision),
        tiempo_promedio: avg(resultadosTable.tiempo_total),
        respuestas_correctas: sum(resultadosTable.respuestas_correctas),
        respuestas_incorrectas: sum(resultadosTable.respuestas_incorrectas),
      })
      .from(resultadosTable)
      .groupBy(resultadosTable.id_usuario)
      .orderBy(desc(sum(resultadosTable.puntuacion)));

    const rankingConPerfiles = await Promise.all(
      rankingData.map(async (entry, index) => {
        const [perfil] = await db
          .select({
            id: perfilesTable.id,
            nombre: perfilesTable.nombre,
            usuario: perfilesTable.usuario,
            rol: perfilesTable.rol,
            grado_bachillerato: perfilesTable.grado_bachillerato,
            avatar_url: perfilesTable.avatar_url,
            puntos_totales: perfilesTable.puntos_totales,
            retos_completados: perfilesTable.retos_completados,
            racha_dias: perfilesTable.racha_dias,
            ultimo_acceso: perfilesTable.ultimo_acceso,
            creado_en: perfilesTable.creado_en,
            ultimo_nivel_intento: perfilesTable.ultimo_nivel_intento,
            mejor_puntaje_por_modulo: perfilesTable.mejor_puntaje_por_modulo,
          })
          .from(perfilesTable)
          .where(eq(perfilesTable.id, entry.id_usuario));

        if (!perfil || perfil.rol !== "estudiante") return null;

        // Filtra por grado de bachillerato que coincide con el modulo
        if (perfil.grado_bachillerato !== moduleId) return null;

        return {
          posicion: index + 1,
          usuario_id: perfil.id,
          nombre: perfil.nombre ?? perfil.usuario,
          usuario: perfil.usuario,
          grado_bachillerato: perfil.grado_bachillerato,
          rol: perfil.rol,
          avatar_url: perfil.avatar_url,
          racha_dias: perfil.racha_dias,
          puntos_totales: Number(entry.puntos_totales || 0),
          retos_completados: Number(entry.retos_completados || 0),
          precision_promedio: Number(entry.precision_promedio || 0),
          tiempo_promedio: Number(entry.tiempo_promedio || 0),
          respuestas_correctas: Number(entry.respuestas_correctas || 0),
          respuestas_incorrectas: Number(entry.respuestas_incorrectas || 0),
        };
      })
    );

    const rankingFiltrado = rankingConPerfiles
      .filter(Boolean)
      .map((entry, index) => ({ ...entry!, posicion: index + 1 }));

    res.json(rankingFiltrado);
  } catch (err) {
    req.log.error({ err }, "Error al obtener ranking por modulo");
    res.status(500).json({ error: "server_error", message: "Error interno" });
  }
});

export default router;

File: artifacts/api-server-python/routers/ranking.py
from fastapi import APIRouter, Depends
from db import get_db, fetchone, fetchall
from auth_utils import get_current_user
from datetime import datetime

router = APIRouter()


def serialize_row(row):
    if row is None:
        return None
    r = dict(row)
    for k, v in r.items():
        if isinstance(v, datetime):
            r[k] = v.isoformat()
    return r


def build_ranking(conn, module_id=None):
    rows = fetchall(conn,
        """SELECT id_usuario,
               SUM(puntuacion) as puntos_totales,
               COUNT(*) as retos_completados,
               AVG(`precision`) as precision_promedio,
               AVG(tiempo_total) as tiempo_promedio,
               SUM(respuestas_correctas) as respuestas_correctas,
               SUM(respuestas_incorrectas) as respuestas_incorrectas
           FROM resultados
           GROUP BY id_usuario
           ORDER BY puntos_totales DESC""")

    ranking = []
    for entry in rows:
        perfil = fetchone(conn,
            "SELECT id, nombre, usuario, rol, grado_bachillerato, avatar_url, puntos_totales, retos_completados, racha_dias, ultimo_acceso, creado_en FROM perfiles WHERE id = %s",
            (entry["id_usuario"],))
        if not perfil or perfil["rol"] != "estudiante":
            continue
        if module_id and perfil.get("grado_bachillerato") != module_id:
            continue
        ranking.append({
            "usuario_id": perfil["id"],
            "nombre": perfil["nombre"] or perfil["usuario"],
            "usuario": perfil["usuario"],
            "grado_bachillerato": perfil.get("grado_bachillerato"),
            "rol": perfil["rol"],
            "avatar_url": perfil.get("avatar_url"),
            "racha_dias": perfil.get("racha_dias"),
            "puntos_totales": float(entry["puntos_totales"] or 0),
            "retos_completados": int(entry["retos_completados"] or 0),
            "precision_promedio": float(entry["precision_promedio"] or 0),
            "tiempo_promedio": float(entry["tiempo_promedio"] or 0),
            "respuestas_correctas": int(entry["respuestas_correctas"] or 0),
            "respuestas_incorrectas": int(entry["respuestas_incorrectas"] or 0),
        })

    return [{"posicion": i + 1, **e} for i, e in enumerate(ranking)]


@router.get("/ranking")
def get_ranking(user=Depends(get_current_user)):
    with get_db() as conn:
        return build_ranking(conn)


@router.get("/ranking/module/{module_id}")
def get_ranking_module(module_id: int, user=Depends(get_current_user)):
    with get_db() as conn:
        return build_ranking(conn, module_id=module_id)
